source/agora/node/Validator.d

```diff
@@ -94,6 +98,9 @@ public class Validator : FullNode, API
         // currently we are not saving preimage info,
         // we only have the commitment in the genesis block
         this.regenerateQuorums(Height(0));
+
+        this.admin_interface = new AdminInterface(config,
+            this.config.validator.key_pair, this.clock);
```

Comment:
What's the point of passing the `config` AND the `key_pair` ?
I had a look at the code, and it is quite tangled:
- If the node needs the full `config` object (which is extremely rare), then it doesn't need to receive `key_pair` as well;
- If the node doesn't need the full `config` object, how can we reduce what we pass.

Looking at the code, it does a bunch of checks that `validator` and `admin` are enabled. This is the wrong way to think about it. If `start` gets called, we just start, without checking the config again.

The important point is to separate concern: You want one part of the code (object, function) to be responsible for only one thing. Duplicating those checks means that responsibility is not clear and things gets tangled (and unworkable) pretty quickly from there.